### PR TITLE
IGNITE-21640 Exchange with ClusterNode instances on handshake

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-21640

### CODE REVIEW

File: modules/network/src/main/java/org/apache/ignite/internal/network/message/ClusterNodeMessage.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import org.apache.ignite.internal.network.NetworkMessage;
+import org.apache.ignite.internal.network.annotations.Transferable;
+import org.apache.ignite.network.ClusterNode;
+import org.apache.ignite.network.NetworkAddress;
+
+/**
+ * {@link ClusterNode} as a network message class.
+ */
+@Transferable(CLUSTER_NODE_MESSAGE)
+public interface ClusterNodeMessage extends NetworkMessage, Serializable {
```
Comment (sashapolo) [inline modules/network/src/main/java/org/apache/ignite/internal/network/message/ClusterNodeMessage.java:33]: We already have a `ClusterNodeMessage`, can we reuse it?

File: modules/network/src/main/java/org/apache/ignite/internal/network/netty/ConnectionManager.java [hunk 2]
```diff
@@ -105,8 +107,7 @@ public class ConnectionManager implements ChannelCreationListener {
     /** Node consistent id. */
     private final String consistentId;
 
-    /** Node launch id. As opposed to {@link #consistentId}, this identifier changes between restarts. */
-    private final UUID launchId;
+    private volatile @Nullable ClusterNode localNode;
```
Comment (sashapolo) [inline modules/network/src/main/java/org/apache/ignite/internal/network/netty/ConnectionManager.java:110]: Why is this field volatile now?

File: modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryClientHandshakeManager.java [hunk 3]
```diff
@@ -237,17 +230,16 @@ private void onHandshakeStartMessage(HandshakeStartMessage handshakeStartMessage
             return;
         }
 
-        this.remoteLaunchId = handshakeStartMessage.launchId();
-        this.remoteConsistentId = handshakeStartMessage.consistentId();
+        this.remoteNode = handshakeStartMessage.serverNode().asClusterNode();
 
-        ChannelKey channelKey = new ChannelKey(remoteConsistentId, remoteLaunchId, connectionId);
+        ChannelKey channelKey = new ChannelKey(remoteNode.name(), UUID.fromString(remoteNode.id()), connectionId);
```
Comment (sashapolo) [inline modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryClientHandshakeManager.java:235]: Should we pass the cluster node to all such places?

File: modules/network/src/main/java/org/apache/ignite/internal/network/message/ClusterNodeMessage.java (hunk 1 above)
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/message/ClusterNodeMessage.java:33]: It's in another module and it is specific to the CMG: take a look at its 3 last methods. I also considered a possibility to create a base interface in the network module and extend it in the CMG modules with those 3 methods, but this does not seem to give a lot of gain (it allows us to reduce duplication by removing 4 lines defining id/name/host/port), but it introduces a dependency which usually bites afterwards. That's why it seemed to me to just copy the interface and remove the 3 last methods from it.

File: modules/network/src/main/java/org/apache/ignite/internal/network/netty/ConnectionManager.java (hunk 2 above)
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/netty/ConnectionManager.java:110]: It is not final. I'm not sure we have a happens-before between the startup of `ConnectionManager` and establishment of connections. Also, this field is used rarely (each time when a connection is established), so the performance penalty should be negligible.

File: modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryClientHandshakeManager.java (hunk 3 above)
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryClientHandshakeManager.java:235]: `ChannelKey` is defined as a triple `<consistentId, launchId, channelIndex>`. According to the `ClusterNodeImpl` code, it considers 2 nodes equal if their name and address match. The ID is not even taken into consideration (which is weird, by the way). So with the current code replacement of `<node, channelIndex>` will just break everything.

Even if it did not break it (due to having a compatible equals semantics), it seems more straightforward and safer to list the components explicitly.

File: modules/network/src/main/java/org/apache/ignite/internal/network/message/ClusterNodeMessage.java (hunk 1 above)
Comment (sashapolo) [reply modules/network/src/main/java/org/apache/ignite/internal/network/message/ClusterNodeMessage.java:33]: Those three last methods are nullable, can we simply pass nulls?

File: modules/network/src/main/java/org/apache/ignite/internal/network/netty/ConnectionManager.java (hunk 2 above)
Comment (sashapolo) [reply modules/network/src/main/java/org/apache/ignite/internal/network/netty/ConnectionManager.java:110]: My question was about why is this field no longer final)

File: modules/network/src/main/java/org/apache/ignite/internal/network/message/ClusterNodeMessage.java (hunk 1 above)
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/message/ClusterNodeMessage.java:33]: This looks like an abuse, to be honest. And that message is in another module.

File: modules/network/src/main/java/org/apache/ignite/internal/network/netty/ConnectionManager.java (hunk 2 above)
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/netty/ConnectionManager.java:110]: Before building that ClusterNode, we need to start the ConnectionManager to obtain a bound server socket from which we take local address, and then apply some transformations to it. The starting is made in the `start()` method. So the field cannot be final.

**Pull request merged**